File: The JSON Format/Java/GUI/api-gui-app/src/main/java/com/dosomedev/data/ApiAccess.java
package com.dosomedev.data;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URI;
import java.net.URISyntaxException;
import java.net.URL;

import com.dosomedev.model.Animal;
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;

public final class ApiAccess {
    private static ApiAccess instance;
    private String apiUri;

    private ApiAccess() {
        this.apiUri = "http://localhost:8080/api";
    }

    public static ApiAccess getInstance() {
        if (instance == null) {
            instance = new ApiAccess();
        }
        return instance;
    }

    public Animal[] getAnimals() {
        Animal[] animals = null;

        try {
            // Prepare API URL.
            URI uri = new URI(this.apiUri + "/animals");
            URL url = uri.toURL();

            // Get JSON from web service.
            HttpURLConnection connection = (HttpURLConnection) url.openConnection();
            connection.setRequestMethod("GET");
            if (connection.getResponseCode() != 200) {
                throw new IOException("Failed to retrieve data from web service.");
            }

            // Convert JSON to object.
            String responseBody = this.loadResponseBody(connection);
            Gson gson = new Gson();
            TypeToken<Animal[]> animalArrayType = new TypeToken<Animal[]>() {};
            animals = gson.fromJson(responseBody, animalArrayType.getType());

        } catch (URISyntaxException ex) {
            System.out.println(ex.getMessage());
        } catch (IOException ex) {
            System.out.println(ex.getMessage());
        }

        return animals;
    }

    private String loadResponseBody(HttpURLConnection connection) {
        StringBuilder responseBody = new StringBuilder();

        try (BufferedReader reader = new BufferedReader(new InputStreamReader(connection.getInputStream()))) {
            String line;
            while ((line = reader.readLine()) != null) {
                responseBody.append(line);
            }
        } catch (IOException ex) {
            System.out.println(ex.getMessage());
        }

        return responseBody.toString();
    }
}

File: The JSON Format/Python/GUI/data/api_access.py
import requests
from model.animal import Animal

class ApiAccess:
    _instance = None

    def __new__(cls):
        if cls._instance is None:
            cls._instance = super().__new__(cls)
        return cls._instance
    
    def __init__(self):
        self.api_uri = "http://localhost:8080/api"

    def get_animals(self):
        animals = []

        # Prepare API URL.
        url = self.api_uri + "/animals"
        try:
            self.response = requests.get(url)

            if self.response.status_code != 200:
                print("Failed to retrieve data from web service.")

            json_data = self.response.json()

            # Convert JSON to array of animals.
            for animal_data in json_data:
                # Extract relevant data.
                animal_type = animal_data.get("animal")
                name = animal_data.get("name")
                breed = animal_data.get("breed")
                age = animal_data.get("age")

                # Create an Animal object and append it to the animals list.
                animal = Animal(animal_type, name, breed, age)
                animals.append(animal)

        except requests.exceptions.RequestException:
            print("The web service could not be reached.")

        return animals
